Model/Service/Sync.php

```diff
@@ -0,0 +1,261 @@
+        }
+        $this->startBenchmark(self::BENCHMARK_SYNC_NAME, self::BENCHMARK_SYNC_BREAKPOINT);
+        $collection->setPageSize(self::API_BATCH_SIZE);
+        $iterator = new Iterator($collection);
+        $iterator->eachBatch(function (NostoIndexCollection $collectionBatch) use ($account, $store) {
+            $this->checkMemoryConsumption('product sync');
+            $op = new UpsertProduct($account, $this->nostoHelperUrl->getActiveDomain($store));
+            $op->setResponseTimeout(self::RESPONSE_TIMEOUT);
+            foreach ($collectionBatch as $productIndex) {
+                if (!$productIndex->getInSync()) {
+                    $op->addProduct($productIndex->getNostoProduct());
+                }
+            }
+            try {
+                $op->upsert();
+                $this->tickBenchmark(self::BENCHMARK_SYNC_NAME);
+            } catch (\Exception $upsertException) {
+                $this->getLogger()->exception($upsertException);
+            } finally {
+                // We will set this as in sync even if there was failures
+                $collectionBatch->markAsInSyncCurrentItemsByStore($store);
+            }
+        });
+        $this->logBenchmarkSummary(self::BENCHMARK_SYNC_NAME, $store);
+        try {
+            $totalDeleted = $this->purgeDeletedProducts($store);
+            $this->getLogger()->info(
+                sprintf(
+                    'Removed total of %d products from index for store %s',
+                    $totalDeleted,
+                    $store->getCode()
+                )
+            );
+        } catch (NostoException $e) {
+            $this->getLogger()->exception($e);
+        }
+    }
+
+    /**
+     * Defines product index as in sync
+     *
+     * @param ProductIndexInterface $productIndex
+     * @throws \Exception
+     * @return void
+     */
+    public function markAsInSync(ProductIndexInterface $productIndex)
+    {
+        if (!$productIndex->getInSync()) {
+            $productIndex->setInSync(true);
+            $this->indexRepository->save($productIndex);
+        }
+    }
+
+    /**
+     * @param int[] $productIds
+     * @param Store $store
+     * @return void
+     */
+    public function markAsInSyncByProductIdsAndStoreId(array $productIds, Store $store)
+    {
+        try {
+            $this->nostoIndexCollectionFactory->create()->markAsInSync($productIds, $store);
+        } catch (\Exception $e) {
+            $this->getLogger()->exception($e);
+        }
+    }
+
+    /**
+     * Discontinues products in Nosto and removes indexed products from Nosto product index
+     *
+     * @param NostoIndexCollection $collection
+     * @param Store $store
+     * @return int number of deleted products
+     * @throws NostoException
+     */
+    public function deleteIndexedProducts(NostoIndexCollection $collection, Store $store)
+    {
+        if ($collection->getSize() === 0) {
```

Comment:
Why is it needed? Is this extraneous size check really important?

Reply:
Let's check this in the iterator